src/daemon/http/request.rs

```diff
@@ -0,0 +1,527 @@
+                } else {
+                    s.to_string()
+                }
+            }),
+        }
+    }
+
+    /// Checks for permissions.
+    ///
+    /// Returns an appropriate error response if the permissions are not met.
+    pub fn check_permission(
+        &self, permission: Permission, resource: Option<&MyHandle>
+    ) -> Result<(), HttpResponse> {
+        self.auth.check_permission(permission, resource).map_err(|err| {
+            HttpResponse::response_from_error(Error::from(err))
+        })
+    }
+
+    /// Checks the permissions and progresses to the next processing stage.
+    ///
+    /// If the authentication information for the request has the given
+    /// permissions, returns an [`AuthedRequest`] and an [`Actor`] which
+    /// allow further processing.
+    ///
+    /// Otherwise returns an appropriate error response.
+    pub fn proceed_permitted(
+        self, 
+        permission: Permission,
+        resource: Option<&MyHandle>,
+    ) -> Result<(AuthedRequest<'a>, AuthInfo), HttpResponse> {
+        self.check_permission(permission, resource)?;
+        Ok((
+            AuthedRequest {
+                request: self.request,
+                server: self.server,
+                limits: self.limits,
+            },
+            self.auth
+        ))
+    }
+
+    /// Permits the request to the next processing stage.
+    ///
+    /// Returns [`AuthedRequest`] and [`Actor`] without requiring any
+    /// permissions whatsoever.
+    pub fn proceed_unchecked(
+        self
+    ) -> (AuthedRequest<'a>, AuthInfo) {
+        (
+            AuthedRequest {
+                request: self.request,
+                server: self.server,
+                limits: self.limits,
+            },
+            self.auth
+        )
+    }
+
+    /// Splits the request into the server and raw Hyper request.
+    pub fn proceed_raw(self) -> (&'a HttpServer, HyperRequest) {
+        (self.server, self.request)
+    }
+}
+
+
+//------------ AuthedRequest -------------------------------------------------
+
+/// A request that has been checked for the correct access permissions.
+///
+/// Only this type allows access to the request’s body and the server.
+pub struct AuthedRequest<'a> {
+    /// The underlying raw request.
+    request: HyperRequest,
+
+    /// The server providing access to Krill itself.
+    server: &'a HttpServer,
+
+    /// The limits for reading the body of the request.
+    limits: BodyLimits,
+}
+
+impl<'a> AuthedRequest<'a> {
+    /// Ensures the body is empty.
+    pub fn empty(self) -> Result<&'a HttpServer, Error> {
+        if self.request.body().size_hint().upper() != Some(0) {
+            // XXX Maybe this should have its own error?
+            return Err(Error::PostTooBig)
+        }
+        Ok(self.server)
+    }
+
+    /// Returns the raw bytes of the request body.
+    pub async fn api_bytes(self) -> Result<(&'a HttpServer, Bytes), Error> {
```

Comment:
Why `api_bytes()` and not `into_bytes()` given that it would work for any request, not just API calls? (aside from the post limit check)

Reply:
Ah I see below `rfc6492_bytes()` ... okay, `into_raw_bytes()` then?

Reply:
The name is just what is was called before. I didn’t find it important enough to rename it. (They are all not great since they aren’t expressing that they actually do something and not just return something they already have.)

Reply:
I have renamed the body reading methods into `read_*`.